synapseclient/core/retry.py

```diff
@@ -6,20 +6,55 @@
 from synapseclient.core.utils import is_json
 from synapseclient.core.dozer import doze
 
+DEFAULT_RETRIES = 3
+DEFAULT_WAIT = 1
+DEFAULT_BACK_OFF = 2
+DEFAULT_MAX_WAIT = 30
+
+DEFAULT_RETRY_STATUS_CODES = [429, 500, 502, 503, 504]
+
+# strings that may appear in responses that suggest a retryable condition
+RETRYABLE_CONNECTION_ERRORS = [
+    "proxy error",
+    "slow down",
+    "timeout",
+    "timed out",
+    "connection reset by peer",
+    "unknown ssl protocol error",
+    "couldn't connect to host",
+    "slowdown",
+    "try again",
+    "connection reset by peer",
+]
+
+# Exceptions that may be retryable. These are socket level exceptions
+# not associated with an HTTP response
+RETRYABLE_CONNECTION_EXCEPTIONS = [
+    "ChunkedEncodingError",
+    "ConnectionError",
+    'ConnectionResetError',
+    "Timeout",
+    "timeout",
+]
+
 
 def with_retry(function, verbose=False,
-               retry_status_codes=[429, 500, 502, 503, 504], retry_errors=[], retry_exceptions=[],
-               retries=3, wait=1, back_off=2, max_wait=30):
+               retry_status_codes=[429, 500, 502, 503, 504], retry_status_code_in=True,
+               retry_errors=[], retry_exceptions=[],
+               retries=DEFAULT_RETRIES, wait=DEFAULT_WAIT, back_off=DEFAULT_BACK_OFF, max_wait=DEFAULT_MAX_WAIT):
     """
     Retries the given function under certain conditions.
 
-    :param function:           A function with no arguments.  If arguments are needed, use a lambda (see example).
-    :param retry_status_codes: What status codes to retry upon in the case of a SynapseHTTPError.
-    :param retry_errors:       What reasons to retry upon, if function().response.json()['reason'] exists.
-    :param retry_exceptions:   What types of exceptions, specified as strings, to retry upon.
-    :param retries:            How many times to retry maximum.
-    :param wait:               How many seconds to wait between retries.
-    :param back_off:           Exponential constant to increase wait for between progressive failures.
+    :param function:                A function with no arguments.  If arguments are needed, use a lambda (see example).
+    :param retry_status_codes:      What status codes to retry upon in the case of a SynapseHTTPError.
+    :param retry_status_code_in:    True if should retry if the status code is in retry_status_codes,
```

Comment:
the name and documentation is slightly confusing. 
maybe `retry_status_code_in` -> `retry_on_status_code_match`?
replace `if should retry` -> ` - retry`
replace `is in` with `exists in`

also, using a string enumeration of retry types might be more clear about the intentions than `True`/`False`
`retry_type` could be : 
`ON_MATCH`  - retry if status code matches values in `retry_status_codes`
`ON_MISMATCH` - retry if status code does not match values in `retry_status_codes`


alternatively, we can use 2 lists:
`retry_status_codes` -  list of status codes to retry if received as a response
`expected_status_codes` - retries until response returns one of the expected codes.